### PR TITLE
fix(memory-policy): normalize finite action mass without overflow

### LINKED ISSUE #314: ExperientialMemoryPolicy sums float32 action mass without overflow protection, incorrectly abstaining on valid retrievals
`ExperientialMemoryPolicy._interpret_jit` directly summed the retrieved float32 categorical action masses via `jnp.sum(safe_finite_mass)`. A valid, finite, nonnegative mass vector such as `[float32_max, float32_max, float32_max]` overflows that sum to `inf`. The policy then set `action_mass_valid=False` and abstained with action `-1`, even though the memory accepted the retrieval and the documented categorical distribution is exactly `[1/3, 1/3, 1/3]` with lowest-index action `0`.

confirmed directly through the real public path (`ExperientialMemory.write` → `.query` → `ExperientialMemoryPolicy.propose`):
```text
wrote True
retrieval.accepted True
retrieval.action [3.4028235e+38 3.4028235e+38 3.4028235e+38]
total_action_mass inf
action_mass_valid False
available False
action -1
normalized [0. 0. 0.]
```

reachability: `ExperientialMemoryPolicy` is exported from both `alberta_framework.core` and the root `alberta_framework` package. `PrototypeAgent` constructs one when experiential memory is enabled (`alberta_framework/core/prototype_agent.py:2321`); its public `scan_transitions` path reaches `policy.propose(...)` with caller-supplied finite float32 action-mass vectors that are not hardcoded to a safe magnitude — the overflow case is reachable with real inputs, not just a synthetic edge.

fix incoming: scale mass by its own max via `frexp`/`ldexp` before summing/normalizing (avoids the intermediate overflow entirely and is scale-invariant, so `normalized_action_mass` is correct regardless of magnitude), and report a finite-saturated float32 total instead of `inf` when the true sum exceeds the representable diagnostic range.

### PR DESCRIPTION
**Pull request by Svector-anu**
Closes #314.

## What changed

`ExperientialMemoryPolicy._interpret_jit` directly summed the retrieved float32 categorical action masses via `jnp.sum(safe_finite_mass)`. A valid, finite, nonnegative mass vector such as `[float32_max, float32_max, float32_max]` overflows that sum to `inf`, so the policy set `action_mass_valid=False` and abstained with action `-1` even though the retrieval was accepted and the documented categorical distribution is exactly `[1/3, 1/3, 1/3]` with lowest-index action `0`.

confirmed directly through the real public path (`ExperientialMemory.write` → `.query` → `ExperientialMemoryPolicy.propose`) before touching the source:

```text
wrote True
retrieval.accepted True
retrieval.action [3.4028235e+38 3.4028235e+38 3.4028235e+38]
total_action_mass inf
action_mass_valid False
available False
action -1
normalized [0. 0. 0.]
```

fix: the policy now scales mass by its own max via `frexp`/`ldexp` before summing/normalizing — this is mathematically scale-invariant (`normalized = mass / sum(mass)` regardless of magnitude) and never overflows the intermediate sum, since every scaled term is bounded in `[0, 1]`. The reported diagnostic `total_action_mass` is finite-saturated at float32 max instead of `inf` when the true sum exceeds the representable range, rather than propagating a `nan`/`inf` sentinel to callers.

verified independently, beyond the report's own test, that the scale-invariant normalization is correct across the full magnitude range (asymmetric normal-scale mass, uniform mass, and an extreme-skew case), not just the uniform-float32_max case the regression test covers:

```text
[2. 1. 1.] -> [0.5  0.25 0.25] sum= 1.0 valid= True
[1. 1. 1.] -> [0.33333334 0.33333334 0.33333334] sum= 1.0 valid= True
[3.4e+38 1.0e+00 1.0e+00] -> [1. 0. 0.] sum= 1.0 valid= True
```
(the last case's `[1, 0, 0]` is float32's own inherent precision loss at a ~1e38 dynamic-range ratio, not a defect introduced by the scaling method — any float32 computation of that ratio underflows the smaller terms to 0.)

No paper formula is involved; the independent reference is first-principles categorical normalization (equal positive masses at any common finite scale have equal probability).

## Reachability

`ExperientialMemoryPolicy` is exported from both `alberta_framework.core` and the root `alberta_framework` package. `PrototypeAgent` constructs one when experiential memory is enabled (`alberta_framework/core/prototype_agent.py:2321`); its public `scan_transitions` path reaches `policy.propose(...)` with caller-supplied finite float32 action-mass vectors that are not hardcoded to a safe magnitude — confirmed directly by grepping the real (non-test) call site before writing this claim.

## Verification

```text
$ .venv/bin/python -m pytest tests/test_experiential_memory_policy.py -q -o addopts=""
29 passed in 90.49s (strict RuntimeWarning mode)

$ .venv/bin/python -m ruff check alberta_framework/core/experiential_memory_policy.py tests/test_experiential_memory_policy.py
All checks passed!

$ .venv/bin/python -m mypy
Success: no issues found in 164 source files
```

New test: `test_finite_action_mass_sum_cannot_overflow_into_false_abstention`, using the real public write/query/propose path with three `float32_max` action masses, asserting the retrieval is accepted, mass is valid, action resolves to lowest-index `0`, `normalized_action_mass` is `[1/3, 1/3, 1/3]`, and `total_action_mass` reports the finite-saturated float32 max instead of `inf`. The pre-existing `test_invalid_or_zero_retrieved_action_mass_fails_closed` parametrization drops the `(3.0e38, 3.0e38, 3.0e38)` case — that mass is no longer invalid, since it was the bug, not a genuinely-invalid input; the negative and all-zero cases correctly remain in the "fails closed" table.

mutation-resistance verified independently: reverted just the source fix (`git stash push -- alberta_framework/core/experiential_memory_policy.py`, kept the test), reran — failed with `assert False` on `action_mass_valid`, reproducing the exact false-abstention behavior described above. Restored the fix, 29/29 green again.

Reviewer follow-up at the final head preserves the legacy direct sum bit-for-bit whenever it is finite, uses a hard finite clamp only after overflow, and retains a seeded 200-vector public-path boundary that failed the prior division-based reconstruction 19 times. The intentional NumPy reference overflow is scoped under `np.errstate`; the full focused file passes with `RuntimeWarning` promoted to an error.

## Evidence

3d216ef14d96350623316f35a317b04167616c39

```text
red (source fix reverted, test kept):
$ .venv/bin/python -m pytest tests/test_experiential_memory_policy.py -q -o addopts="" -k test_finite_action_mass_sum_cannot_overflow_into_false_abstention
FAILED tests/test_experiential_memory_policy.py::test_finite_action_mass_sum_cannot_overflow_into_false_abstention - assert False
1 failed, 27 deselected in 1.10s

green (fix restored):
$ .venv/bin/python -m pytest tests/test_experiential_memory_policy.py -q -o addopts=""
29 passed in 90.49s (strict RuntimeWarning mode)
```

```text
$ .venv/bin/python -c "
import jax.numpy as jnp
_FLOAT32_MAX = 3.4028234663852886e38
def normalize(safe_finite_mass):
    mass_scale = jnp.max(safe_finite_mass)
    scale_mantissa, scale_exponent = jnp.frexp(mass_scale)
    safe_mantissa = jnp.where(mass_scale > 0.0, scale_mantissa, 1.0)
    scaled_mass = jnp.ldexp(safe_finite_mass, -scale_exponent) / safe_mantissa
    scaled_total = jnp.sum(scaled_mass)
    valid = scaled_total > 0.0
    safe_scaled_total = jnp.where(valid, scaled_total, 1.0)
    normalized = jnp.where(valid, scaled_mass / safe_scaled_total, jnp.zeros_like(safe_finite_mass))
    return normalized, scaled_total, valid
for case in [jnp.array([2.0, 1.0, 1.0], dtype=jnp.float32), jnp.array([1.0,1.0,1.0], dtype=jnp.float32), jnp.array([3.4e38, 1.0, 1.0], dtype=jnp.float32)]:
    norm, total, valid = normalize(case)
    print(case, '->', norm, 'sum=', float(jnp.sum(norm)), 'valid=', bool(valid))
"
[2. 1. 1.] -> [0.5  0.25 0.25] sum= 1.0 valid= True
[1. 1. 1.] -> [0.33333334 0.33333334 0.33333334] sum= 1.0 valid= True
[3.4e+38 1.0e+00 1.0e+00] -> [1. 0. 0.] sum= 1.0 valid= True
```
independently reran the focused suite, ruff, and mypy myself; independently verified the normalization stays correctly scale-invariant (sums to 1) across normal, uniform, and extreme-skew magnitudes, not just the uniform-max case in the regression test; independently confirmed the export chain and real call site before writing the reachability claim; did my own revert-and-confirm-red mutation check.

## Provenance

AI provider/model: openai / gpt-5.6-sol (fix, tests, via AgentRouter proxy); anthropic / claude-sonnet-5 (independent re-verification: reran the full test file, ruff, and mypy myself, independently probed the scale-invariant normalization math across additional magnitude cases beyond the report's own test, verified the export chain and real call site directly, did my own revert-and-confirm-red mutation check, opened the governing issue, pushed, opened this PR)
Client / agent tooling: codex via AgentRouter proxy; Claude Code
Attribution status: self-reported

### CODE REVIEW

**Review by lalalune — CHANGES_REQUESTED** [review]
Blocking exact-head numerical finding at 57398e9864294fd0795e1af5e0b8fdcbfeb5a369: the advertised finite saturation is not closed. `max_float32 / scaled_total` can round upward, so multiplying the resulting scale by `scaled_total` overflows back to `inf`; an independent 500,000-vector positive-finite float32 fuzz found 11 nonfinite `total_action_mass` outputs. The new recomputation also changes previously finite mass bit patterns broadly instead of preserving the direct finite sum. Please use a saturation construction that is provably finite after float32 rounding, preserve the legacy direct sum whenever it is already finite, add a retained public-path regression/fuzz boundary, and rerun focused/adjacent/static/fresh CI.

**Comment by Svector-anu** [comment]
Thanks — this was a real gap, appreciated the precision of the fuzz finding.

Addressed all four points:

1. **Provably-finite saturation construction.** Dropped the division-based reconstruction (`_FLOAT32_MAX / scaled_total`, then multiply back — which can round upward past `_FLOAT32_MAX`). Now: `jnp.minimum(mass_scale * safe_scaled_total, _FLOAT32_MAX)`. Even when the multiplication itself overflows to `inf` in float32 (which it still can, e.g. at the exact reported bug case), `jnp.minimum(inf, _FLOAT32_MAX)` correctly returns `_FLOAT32_MAX` — the clamp is unconditional and doesn't depend on the intermediate arithmetic staying finite.

2. **Preserve the legacy direct sum whenever it's already finite.** `total_action_mass` now uses `jnp.sum(safe_finite_mass)` (the original, pre-my-PR computation) directly whenever that's finite, and only falls back to the scaled/clamped reconstruction when the direct sum overflows. So for every normal-magnitude input, the output is bit-identical to what `main` produced before my PR touched this at all.

3. **Retained regression/fuzz boundary on the public path.** Added `test_large_finite_action_mass_total_never_overflows_on_rounding`: 200 iterations of random positive-finite float32 triples through the real `ExperientialMemory.write` → `.query` → `ExperientialMemoryPolicy.propose` path, asserting `total_action_mass` is always finite, and equals the direct sum whenever that's finite. Independently confirmed this test actually catches your finding — reverted just to the old division-based reconstruction, reran, and it fails on a concrete case: `total_action_mass=inf` for masses `(8.226879994999578e+37, 1.0839146855179014e+38, 3.2805915249088818e+38)`. Restored the fix, green again.

4. **Reran focused/adjacent/static/fresh CI:**

```text
$ .venv/bin/python -m pytest tests/test_experiential_memory_policy.py -q -o addopts=""
29 passed in 7.26s

$ .venv/bin/python -m pytest tests/ -q -o addopts="" -k experiential_memory
76 passed, 1 skipped, 6905 deselected in 21.91s

$ .venv/bin/python -m ruff check alberta_framework/core/experiential_memory_policy.py tests/test_experiential_memory_policy.py
All checks passed!

$ .venv/bin/python -m mypy
Success: no issues found in 164 source files
```

Also independently reran your own fuzz shape at a larger scale before pushing (500k random positive-finite float32 triples via a standalone extraction of the same math): old formula produced 48,521 non-finite outputs, new formula produced 0, with 0 mismatches against the legacy direct sum wherever it was already finite.

head: `62501684fea5c20c41774f75e3ca269e803b58c9`

**Review by lalalune — APPROVED** [review]
Repaired exact head 3d216ef independently revalidated: the old finite-saturation counterexample is closed, legacy finite totals are preserved, the seeded boundary catches the old formula, and the focused suite is warning-clean under RuntimeWarning-as-error (29 passed). Ruff and targeted strict mypy pass. Approved pending the newly authorized full 20-job CI matrix; do not merge before it is green.

**Review by lalalune — APPROVED** [review]
Independent exact-head review at 3d216ef14d96350623316f35a317b04167616c39 found the reachable overflow repair sound. It preserves the direct finite total_action_mass bits and clamps only the overflowing diagnostic with a one-ulp-safe finite bound; the public adversarial vector now lands at float32 max. Exact 29 tests under warnings-as-errors, repo Ruff, strict mypy, and current-main composition are green. The upstream subnormal weighting loss occurs before this policy receives retrievals and is not introduced by this PR. Merge only after the exact head reaches 20/20 CI.
